# Hasan Mahabub Sayem

# Missing value handling Assignment 10

titanic dataset: https://www.kaggle.com/competitions/titanic

- create new column (column name 'initials') from Name column using only initials (mr./mrs./cap. etc are examples of initials and they end with a full stop ) 
- then fill missing Age value base on 'initials' column (using groupby mean)

In [2]:
import pandas as pd
import numpy as np

In [121]:
df = pd.read_csv('./data/train.csv')
df.sample(3)

PassengerId  Survived  Pclass                                       Name  \
763          764         1       1  Carter, Mrs. William Ernest (Lucile Polk)   
377          378         0       1                  Widener, Mr. Harry Elkins   
380          381         1       1                      Bidois, Miss. Rosalie   

        Sex   Age  SibSp  Parch    Ticket     Fare    Cabin Embarked  
763  female  36.0      1      2    113760  120.000  B96 B98        S  
377    male  27.0      0      2    113503  211.500      C82        C  
380  female  42.0      0      0  PC 17757  227.525      NaN        C

In [87]:
df.shape

(891, 12)

In [122]:
# removing duplicate value if any
df.drop_duplicates()

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [98]:
# calculating percentage of missing value
(df.isnull().sum() / df.shape[0]) * 100

PassengerId     0.000000
Survived        0.000000
Pclass          0.000000
Name            0.000000
Sex             0.000000
Age            19.865320
SibSp           0.000000
Parch           0.000000
Ticket          0.000000
Fare            0.000000
Cabin          77.104377
Embarked        0.224467
dtype: float64

In [72]:
# getting initials from name column
def initial(row):        
    row['initials'] = [a for a in row.Name.split(' ') if a.endswith('.')][0]
    return row
new_df = df.apply(initial,  axis='columns')
new_df.sample(5)

PassengerId  Survived  Pclass                                      Name  \
188          189         0       3                          Bourke, Mr. John   
280          281         0       3                          Duane, Mr. Frank   
699          700         0       3  Humblen, Mr. Adolf Mathias Nicolai Olsen   
389          390         1       2                     Lehmann, Miss. Bertha   
162          163         0       3                Bengtsson, Mr. John Viktor   

        Sex   Age  SibSp  Parch   Ticket    Fare  Cabin Embarked initials  
188    male  40.0      1      1   364849  15.500    NaN        Q      Mr.  
280    male  65.0      0      0   336439   7.750    NaN        Q      Mr.  
699    male  42.0      0      0   348121   7.650  F G63        S      Mr.  
389  female  17.0      0      0  SC 1748  12.000    NaN        C    Miss.  
162    male  26.0      0      0   347068   7.775    NaN        S      Mr.

In [76]:
# reading groupwise mean age value
new_df.groupby('initials')['Age'].mean()

initials
Capt.        70.000000
Col.         58.000000
Countess.    33.000000
Don.         40.000000
Dr.          42.000000
Jonkheer.    38.000000
Lady.        48.000000
Major.       48.500000
Master.       4.574167
Miss.        21.773973
Mlle.        24.000000
Mme.         24.000000
Mr.          32.368090
Mrs.         35.898148
Ms.          28.000000
Rev.         43.166667
Sir.         49.000000
Name: Age, dtype: float64

In [99]:
new_df['Age'] = new_df.groupby('initials')['Age'].transform(lambda x: x.fillna(x.mean()))

PassengerId  Survived  Pclass                                      Name  \
833          834         0       3                    Augustsson, Mr. Albert   
325          326         1       1                  Young, Miss. Marie Grice   
699          700         0       3  Humblen, Mr. Adolf Mathias Nicolai Olsen   
490          491         0       3      Hagland, Mr. Konrad Mathias Reiersen   
147          148         0       3          Ford, Miss. Robina Maggie "Ruby"   

        Sex       Age  SibSp  Parch      Ticket      Fare  Cabin Embarked  \
833    male  23.00000      0      0      347468    7.8542    NaN        S   
325  female  36.00000      0      0    PC 17760  135.6333    C32        C   
699    male  42.00000      0      0      348121    7.6500  F G63        S   
490    male  32.36809      1      0       65304   19.9667    NaN        S   
147  female   9.00000      2      2  W./C. 6608   34.3750    NaN        S   

    initials  
833      Mr.  
325    Miss.  
699      Mr.  
490      Mr.  
147    Miss.

In [120]:
new_df[df.Age.isnull()]

PassengerId  Survived  Pclass                                      Name  \
5              6         0       3                          Moran, Mr. James   
17            18         1       2              Williams, Mr. Charles Eugene   
19            20         1       3                   Masselmani, Mrs. Fatima   
26            27         0       3                   Emir, Mr. Farred Chehab   
28            29         1       3             O'Dwyer, Miss. Ellen "Nellie"   
..           ...       ...     ...                                       ...   
859          860         0       3                          Razi, Mr. Raihed   
863          864         0       3         Sage, Miss. Dorothy Edith "Dolly"   
868          869         0       3               van Melkebeke, Mr. Philemon   
878          879         0       3                        Laleff, Mr. Kristo   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   

        Sex        Age  SibSp  Parch      Ticket     Fare Cabin Embarked  \
5      male  32.368090      0      0      330877   8.4583   NaN        Q   
17     male  32.368090      0      0      244373  13.0000   NaN        S   
19   female  35.898148      0      0        2649   7.2250   NaN        C   
26     male  32.368090      0      0        2631   7.2250   NaN        C   
28   female  21.773973      0      0      330959   7.8792   NaN        Q   
..      ...        ...    ...    ...         ...      ...   ...      ...   
859    male  32.368090      0      0        2629   7.2292   NaN        C   
863  female  21.773973      8      2    CA. 2343  69.5500   NaN        S   
868    male  32.368090      0      0      345777   9.5000   NaN        S   
878    male  32.368090      0      0      349217   7.8958   NaN        S   
888  female  21.773973      1      2  W./C. 6607  23.4500   NaN        S   

    initials  
5        Mr.  
17       Mr.  
19      Mrs.  
26       Mr.  
28     Miss.  
..       ...  
859      Mr.  
863    Miss.  
868      Mr.  
878      Mr.  
888    Miss.  

[177 rows x 13 columns]

177 age row populated with thier respective intials mean age